smithy-cli/src/main/java/software/amazon/smithy/cli/commands/UpgradeCommand.java

```diff
@@ -0,0 +1,522 @@
+        }
+    }
+
+    String upgradeFile(Model completeModel, Path filePath) {
+        String contents = IoUtils.readUtf8File(filePath);
+        if (VERSION_2.matcher(contents).find()) {
+            return contents;
+        }
+
+        ShapeUpgradeVisitor visitor = new ShapeUpgradeVisitor(completeModel, contents);
+
+        completeModel.shapes()
+                .filter(shape -> shape.getSourceLocation().getFilename().equals(filePath.toString()))
+                // Apply updates to the shapes at the bottom of the file first.
+                // This lets us modify the file without invalidating the existing
+                // source locations.
+                .sorted(new SourceLocationSorter().reversed())
+                .forEach(shape -> shape.accept(visitor));
+
+        return updateVersion(visitor.getModelString());
+    }
+
+    private String updateVersion(String modelString) {
+        Matcher matcher = VERSION_1.matcher(modelString);
+        if (matcher.find()) {
+            return matcher.replaceFirst(format("\\$version: \"2.0\"%n"));
+        }
+        return format("$version: \"2.0\"%n%n") + modelString;
+    }
+
+    private static final class ResultConsumer implements Consumer<ProjectionResult>, BiConsumer<String, Throwable> {
+        private Throwable error;
+        private ProjectionResult result;
+
+        @Override
+        public void accept(String name, Throwable throwable) {
+            this.error = throwable;
+        }
+
+        @Override
+        public void accept(ProjectionResult projectionResult) {
+            // We only expect one result because we're only building one
+            // projection - the source projection.
+            this.result = projectionResult;
+        }
+
+        ProjectionResult getResult() {
+            if (error != null) {
+                throw new RuntimeException(error);
+            }
+            return result;
+        }
+    }
+
+    // Sorts shapes in the order they appear in the file.
+    private static class SourceLocationSorter implements Comparator<FromSourceLocation>, Serializable {
+        @Override
+        public int compare(FromSourceLocation s1, FromSourceLocation s2) {
+            SourceLocation sourceLocation = s1.getSourceLocation();
+            SourceLocation otherSourceLocation = s2.getSourceLocation();
+
+            if (!sourceLocation.getFilename().equals(otherSourceLocation.getFilename())) {
+                return sourceLocation.getFilename().compareTo(otherSourceLocation.getFilename());
+            }
+
+            int lineComparison = Integer.compare(sourceLocation.getLine(), otherSourceLocation.getLine());
+            if (lineComparison != 0) {
+                return lineComparison;
+            }
+
+            return Integer.compare(sourceLocation.getColumn(), otherSourceLocation.getColumn());
+        }
+    }
+
+    private static class TraitSorter implements Comparator<Trait>, Serializable {
```

Comment:
This can also use the Sorter in https://github.com/awslabs/smithy/pull/1192 since Traits also implements FromSourceLocation.